CRM/Civicaseextras/CaseDurationLog.php

```diff
@@ -0,0 +1,388 @@
+      WHERE case_id = %3
+      AND end_task IS NULL
+    ";
+    $params = array(
+      1 => array($activityID, 'Integer'),
+      2 => array($endDate, 'String'),
+      3 => array($caseID, 'Integer'),
+    );
+    CRM_Core_DAO::executeQuery($query, $params);
+  }
+
+  /**
+   * Uses data of the provided case activity object to check if the case is
+   * being opened.
+   *
+   * @param CRM_Core_DAO $caseActivity
+   *   Object with the data for the case and activity.
+   *
+   * @return bool
+   *   True if the current activity is opening the case, false otherwise
+   */
+  private function wasCaseOpening($caseActivity) {
+    $changeStatusToOpen = FALSE;
+    $activityType = CRM_Utils_Array::value($caseActivity->activity_type_id, $this->activityTypes);
+
+    if ($activityType === 'Change Case Status') {
+      if ($this->extractStatusFromSubject($caseActivity->subject) === 'Opened') {
+        $changeStatusToOpen = TRUE;
+      }
+    }
+
+    if ($activityType === 'Open Case' || $changeStatusToOpen) {
+      $query = "
+        SELECT *
+        FROM civicrm_case_duration_log
+        WHERE case_id = %1
+        AND end_task IS NULL
+      ";
+      $dbResult = CRM_Core_DAO::executeQuery($query, array(
+        1 => array($caseActivity->case_id, 'Integer')
+      ));
+
+      if ($dbResult->N === 0) {
+        return TRUE;
+      }
+    }
+
+    return FALSE;
+  }
+
+  /**
+   * Checks if the given case activity is closing the case.
+   *
+   * @param CRM_Core_DAO $caseActivity
+   *   Object with the data for the case and activity.
+   *
+   * @return bool
+   *   True if the current activity is closing the case, false otherwise
+   */
+  private function wasCaseClosing($caseActivity) {
+    $activityType = CRM_Utils_Array::value($caseActivity->activity_type_id, $this->activityTypes);
+
+    if ($activityType === 'Change Case Status') {
+      if ($this->extractStatusFromSubject($caseActivity->subject) === 'Closed') {
+        return TRUE;
+      }
+    }
+
+    return FALSE;
+  }
+
+  /**
+   * Given an activity's subject, it tries to extract from the new case status,
+   * if it was opened or closed.
+   *
+   * @param string $activitySubject
+   *
+   * @return string|null
+   *   Returns 'Opened' or 'Closed' if it is able to extract the case status from
+   *   the subject. Otherwise it returns null.
+   */
+  private function extractStatusFromSubject($activitySubject) {
+    if (stripos($activitySubject, 'Case status changed from') !== FALSE) {
+      $subjectData = explode(' ', $activitySubject);
+      $newCaseStatus = $subjectData[sizeof($subjectData) - 1];
+
+      if (isset($this->statusGroupingsPerLabel[$newCaseStatus])) {
+        return $this->statusGroupingsPerLabel[$newCaseStatus];
+      }
+    }
+
+    return NULL;
+  }
+
+  /**
+   * Calculates the case's duration and stores it in the duration custom field.
+   *
+   * @param $caseID
+   */
+  public function calculateCaseDuration($caseID) {

```

Comment:
As this class is already pretty big maybe a separate `CaseDurationCalculator` would be useful?

Also I think the name `calculate` is misleading since it actually sets the case duration in the last line. Maybe a separate updating it and calculating it?